mcts.py

```diff
@@ -51,34 +14,116 @@ def exploration_bonus(edge, c_puct):
+        how good the edge is to explore with our exploration rate.
     """
-    def score(edge, c_puct):
-        return edge.mean_action_value + exploration_bonus(edge, c_puct)
-    index = np.argmax([score(edge, c_puct) for edge in node.outgoing_edges])
+    def score(edge):
+        return edge.mean_action_value + exploration_bonus(edge)
+    index = np.argmax([score(edge) for edge in node.outgoing_edges])
     return node.outgoing_edges[index]
 
 
-def evaluate(node, model, env):
-    """
-    Return the probability vector for actions,
-    and the value of the current state
+def backup(node, value):
     """
-    prob_vector, value = model(node.state, env)
-    return prob_vector, value
-
-
-def backprop(node, value):
-    """
-    Propagate the value for the current node back
+    Propagate the value for the current node back up
+    the tree
     """
     cur_node = node
     # while not root, move the value up
     while cur_node.in_edge is not None:
-        edge = node.in_edge
+        edge = cur_node.in_edge
         edge.num_visits += 1
         edge.total_action_value += value
         cur_node = edge.in_node
+
+
+def expand_node(node, model, env):
+    action_probs, value = model(node)
+    for i, action_prob in enumerate(action_probs):
+        action = i  # actions are just indexes
+        next_state = env.get_next_state(node.state, action)
+        child_node = Node(next_state)
+        create_new_connection(node, child_node, action, action_prob)
+    return value
+
+
+def perform_rollouts(root_node,
+                     n_leaf_expansions,
+                     model,
+                     env,
+                     exploration_bonus):
+    """
+    Parameters
+    ----------
+    root_node: Node
+        initial node to start MCTS
+    n_leaf_expansions: int
+        number of leaves to expand in each iteration of MCTS when picking an action
+    model: function
+        Model to use for computing the value of each state,
+        prob_vector, value = model(node.state, env)
+    env:
+        game playing environment that can progress game state and give us legal moves
+    exploration_bonus: function
+        Function that takes (edge) as input and returns a score based on
+        how good the edge is to explore with our exploration rate.
+    """
+    cur_node = root_node
+    # add all edges and children for current node
+    value = expand_node(root_node, model, env)
+    while n_leaf_expansions > 0:
+        # expand root
+        edge = select(root_node, exploration_bonus)
+        while edge.num_visits != 0:
+            cur_node = edge.out_node
+            edge = select(cur_node, exploration_bonus)
+        cur_node = edge.out_node
+        # find a node you haven't expanded yet, expand it
+        value = expand_node(cur_node, model, env)
+        backup(cur_node, value)
+
+        n_leaf_expansions -= 1
+
+
+def get_action_distribution(start_state,
```

Comment:
Have we written the code that actually selects from this distribution yet? No right?

Reply:
Correct. I think we still need to write code that will actually run n game of self play, with actions elected using MCTS, and then generate the resulting (state, action_distribution, winner=+/- 1) dataset

Just to make sure, @rajivsambharya you haven't already done that right? I can add that code if not.

Reply:
Yeah, I'm working on that now lol!  I'll take care of this.